server/appenders/cassandra/appender/src/main/java/org/kaaproject/kaa/server/appenders/cassandra/appender/CassandraLogEventDao.java

```diff
@@ -0,0 +1,212 @@
+
+    @Override
+    public void createTable(String tableName) {
+        LOG.info("Create table {} in cassandra keyspace {}", tableName, keyspaceName);
+        String createTableQuery = CREATE_TABLE.replace(TABLE_NAME, tableName);
+        createTableQuery = createTableQuery.replace(KEYSPACE_NAME, keyspaceName);
+        LOG.debug("Execute query {}", createTableQuery);
+        getSession().execute(createTableQuery);
+    }
+
+    @Override
+    public List<LogEventDto> save(List<LogEventDto> logEventDtoList, String tableName) {
+        LOG.debug("Execute bath request for cassandra table {}", tableName);
+        executeBatch(prepareQuery(logEventDtoList, tableName));
+        return logEventDtoList;
+    }
+
+    @Override
+    public ListenableFuture<ResultSet> saveAsync(List<LogEventDto> logEventDtoList, String tableName) {
+        LOG.debug("Execute async bath request for cassandra table {}", tableName);
+        return executeBatchAsync(prepareQuery(logEventDtoList, tableName));
+    }
+
+    @Override
+    public void removeAll(String tableName) {
+        LOG.info("Truncate all data from table {}", tableName);
+        getSession().execute(QueryBuilder.truncate(keyspaceName, tableName));
+    }
+
+    @Override
+    public void close() {
+        LOG.info("Close connection to cassandra cluster.");
+        if (cluster != null) {
+            cluster.close();
+        }
+    }
+
+    private Session getSession() {
+        if (session == null || session.isClosed()) {
+            session = cluster.newSession();
+        }
+        return session;
+    }
+
+    private ConsistencyLevel getWriteConsistencyLevel() {
+        if (writeConsistencyLevel == null) {
+            writeConsistencyLevel = ConsistencyLevel.ONE;
+        }
+        return writeConsistencyLevel;
+    }
+
+    private void executeBatch(RegularStatement... statement) {
+        getSession().execute(prepareBatch(statement));
+    }
+
+    private ResultSetFuture executeBatchAsync(RegularStatement... statement) {
+        return getSession().executeAsync(prepareBatch(statement));
+    }
+
+    private String getId(String id) {
+        if (id == null || id.length() == 0) {
+            id = new UUID(System.currentTimeMillis(), RANDOM.nextLong()).toString();
+        }
+        return id;
+    }
+
+    private Batch prepareBatch(RegularStatement... statement) {
+        Batch batch;
+        if (batchType != null && batchType.equals(CassandraBatchType.UNLOGGED)) {
+            batch = QueryBuilder.unloggedBatch(statement);
+        } else {
+            batch = QueryBuilder.batch(statement);
+        }
+        batch.setConsistencyLevel(getWriteConsistencyLevel());
+        return batch;
+    }
+
+    private Insert[] prepareQuery(List<LogEventDto> logEventDtoList, String collectionName) {
+        Insert[] insertArray = new Insert[logEventDtoList.size()];
+        for (int i = 0; i < logEventDtoList.size(); i++) {
+            LogEventDto dto = logEventDtoList.get(i);
```

Comment:
This may be inefficient in case of LinkedList, etc.. Maybe we should loop and return List instead of array?